controllers/hostingautoscalingpolicy/hostingautoscalingpolicy_controller.go

```diff
@@ -0,0 +1,450 @@
+	}
+	ctx.Log.Info("Determined action for AutoscalingJob", "action", action)
+
+	// If update or delete, delete the existing Policy.
+	if action == controllers.NeedsDelete || action == controllers.NeedsUpdate {
+		if err = r.deleteAutoscalingPolicy(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to delete HostingAutoscalingPolicy"))
+		}
+
+		// Delete succeeded, set Description to nil.
+		ctx.ScalingPolicyDescriptionList = nil
+		ctx.ScalableTargetDescriptionList = nil
+	}
+
+	// If update or create, create the desired HAP.
+	if action == controllers.NeedsCreate || action == controllers.NeedsUpdate {
+		if ctx.ScalableTargetDescriptionList, ctx.ScalingPolicyDescriptionList, err = r.applyAutoscalingPolicy(ctx); err != nil {
+			return r.updateStatusAndReturnError(ctx, FailedAutoscalingJobStatus, errors.Wrap(err, "Unable to apply HostingAutoscalingPolicy"))
+		}
+	}
+
+	// Update the status accordingly.
+	// Review: Deleted Status may be unnecessary
+	status := CreatedAutoscalingJobStatus
+	if ctx.ScalingPolicyDescriptionList == nil {
+		status = DeletedAutoscalingJobStatus
+	}
+	if err = r.updateStatus(ctx, status); err != nil {
+		return err
+	}
+
+	// Remove the Finalizer on delete
+	if controllers.HasDeletionTimestamp(ctx.HostingAutoscalingPolicy.ObjectMeta) {
+		if err = r.removeFinalizer(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Removes the finalizer held by our controller.
+func (r *Reconciler) removeFinalizer(ctx reconcileRequestContext) error {
+	var err error
+
+	ctx.HostingAutoscalingPolicy.ObjectMeta.Finalizers = controllers.RemoveString(ctx.HostingAutoscalingPolicy.ObjectMeta.Finalizers, controllers.SageMakerResourceFinalizerName)
+	if err = r.Update(ctx, ctx.HostingAutoscalingPolicy); err != nil {
+		return errors.Wrap(err, "Failed to remove finalizer")
+	}
+	ctx.Log.Info("Finalizer has been removed")
+
+	return nil
+}
+
+// getResourceIDListfromInputSpec converts the list of resources into a string list to be used for various API calls
+func (r *Reconciler) getResourceIDListfromInputSpec(ctx *reconcileRequestContext) error {
+
+	resourceIDListfromSpec := ctx.HostingAutoscalingPolicy.Spec.ResourceID
+	for _, resourceIDfromSpec := range resourceIDListfromSpec {
+		resourceID := sdkutil.ConvertAutoscalingResourceToString(*resourceIDfromSpec)
+		ctx.ResourceIDList = append(ctx.ResourceIDList, *resourceID)
+	}
+
+	return nil
+}
+
+// Initialize fields on the context object which will be used later.
+func (r *Reconciler) initializeContext(ctx *reconcileRequestContext) error {
+	var err error
+
+	// Ensure we are using the job name specified in the spec
+	if ctx.HostingAutoscalingPolicy.Spec.PolicyName != nil && len(*ctx.HostingAutoscalingPolicy.Spec.PolicyName) > 0 {
+		ctx.PolicyName = *ctx.HostingAutoscalingPolicy.Spec.PolicyName
+	} else {
+		ctx.PolicyName = DefaultAutoscalingPolicyName
```

Comment:
Is it possible this policy would conflict with another? We don't need to autogenerate the policy name?

Reply:
To the best of my tests no generation is not needed. The unique identifier here is the resource ID which means that a single variant cannot have more than one policy. Different variants can have the same policy name and i guess the most standard name is the default one `SageMakerEndpointInvocationScalingPolicy`. 

That being said, i have limited insight into the custom policy name. I did discuss this with Gautam and Michael and this seems ok.